src/analysis_and_optimization/Optimize.ml

```diff
@@ -1017,65 +1017,224 @@ let block_fixing mir =
+    Monotone_framework.minimal_variables_mfp
+      (module Fwd_Flowgraph)
+      (module Rev_Flowgraph)
+      flowgraph_to_mir initial_variables gen_variables
+  in
+  let optimize_min_vars_stmt_base i stmt_pattern =
+    let variable_set = (Map.find_exn soa_levels i).exit in
+    let stmt_val =
+      Stmt.Fixed.Pattern.map (update_expr variable_set)
+        (fun x -> x)
+        stmt_pattern
+    in
+    update_stmt stmt_val variable_set
+  in
+  map_rec_stmt_loc_num flowgraph_to_mir optimize_min_vars_stmt_base
+    (Map.find_exn flowgraph_to_mir 1)
+
+(**
+ * The gen function for the monotone framework for deducing
+ *  the statements and expressions that must be demoted to AoS.
+ * Returns a Set of strings where the strings are the names of 
+ * variables that must be demoted from SoA -> AoS.
+ *)
+let gen_aos_variables
+    (flowgraph_to_mir : (int, Stmt.Located.Non_recursive.t) Map.Poly.t)
+    (l : int) (aos_variables : string Set.Poly.t) =
+  let mir_node mir_idx = Map.find_exn flowgraph_to_mir mir_idx in
+  match (mir_node l).pattern with stmt ->
+    Mem_pattern.query_demotable_stmt aos_variables stmt
+
+(**
+* Deduces whether types can be Structures of Arrays (SoA/fast) or
+*  Arrays of Structs (AoS/slow). See the docs in
+*  Mem_pattern.query_demote_stmt/exprs* functions for 
+*  details on the rules surrounding when demotion from
+*  SoA -> AoS needs to happen.
+*
+* This first does a simple iter over
+* the log_prob portion of the MIR, finding the names of all matrices
+* (and arrays of matrices) where either the Stan math function
+* does not support SoA or is the object is single cell accesed within a 
+* For or While loop. These are the initial variables
+* given to the monotone framework. Then log_prob has all matrix like objects
+* and the functions that use them to SoA. After that the 
+* Monotone framework is used to deduce assignment paths of AoS -> SoA
+* and vice versa which need to be demoted to AoS as well as updating 
+* functions and objects after these assignment passes that then 
+* also need to be AoS.
+*
+* @param mir: The program's whole MIR.
+*)
+let optimize_soa (mir : Program.Typed.t) =
+  let initial_variables =
+    Set.Poly.union_list
+      (List.map
+         ~f:(Mem_pattern.query_initial_demotable_stmt false)
+         mir.log_prob)
+  in
+  let all_eigen_types =
+    Set.Poly.union_list (List.map ~f:Mem_pattern.get_eigen_decls mir.log_prob)
+  in
+  let promotable_types = Set.Poly.diff all_eigen_types initial_variables in
+  let promote_stmt stmt =
+    Mem_pattern.modify_stmt Common.Helpers.SoA stmt promotable_types
+  in
+  let forced_logprob = List.map ~f:promote_stmt mir.log_prob in
+  let demote_exprs aos_exits =
+    Mir_utils.map_rec_expr
+      (Mem_pattern.modify_expr_pattern Common.Helpers.AoS aos_exits)
+  in
+  let modify_stmt_pattern =
+    Mem_pattern.modify_stmt_pattern Common.Helpers.AoS
+  in
+  let transform stmt =
+    optimize_mem_pattern ~gen_variables:gen_aos_variables
```

Comment:
My hope in writing optimize_minimal_variables was that we could use for both optimize_ad_levels and optimize_soa. I think as it stands now, optimize_mem_pattern is almost identical to optimize_minimal_variables except that it's slightly more general. Can we combine those two by renaming optimize_mem_pattern to optimize_minimal_variables, and then changing optimize_ad_levels to use that more general interface?

Reply:
Yes!

Reply:
Managed to sort this out so they both use `optimize_minimal_variables`